File: filter.go
package main

import (
	"io"
	"log"
	"os"
	"os/exec"
	"sort"
	"strings"
)

func choose(strs []string, f func(string) bool) []string {
	var r []string
	for _, str := range strs {
		if f(str) {
			r = append(r, str)
		}
	}
	return r
}

func contains(strs []string, str string) bool {
	for _, s := range strs {
		if s == str {
			return true
		}
	}
	return false
}

func uniq(strs []string) []string {
	var r []string
	for _, str := range strs {
		if !contains(r, str) {
			r = append(r, str)
		}
	}
	return r
}

type byLength []string

func (s byLength) Len() int {
	return len(s)
}
func (s byLength) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}
func (s byLength) Less(i, j int) bool {
	return len(s[i]) < len(s[j])
}

func unescape(str string) string {
	out, err := exec.Command("sh", "-c", "echo "+str).Output()
	if err != nil {
		log.Fatal(err)
	}
	return strings.TrimRight(string(out), "\r\n")
}

func mapStrs(strs []string, f func(str string) string) []string {
	var r []string
	for _, str := range strs {
		r = append(r, f(str))
	}
	return r
}

type readable interface {
	read() ([]byte, error)
}

type stdin struct {
	in io.Reader
}

func (r stdin) read() ([]byte, error) {
	char := make([]byte, 1)
	_, err := r.in.Read(char)
	return char, err
}

type filter struct {
	in   readable
	str  string
	buf  []byte
	mask []byte
}

func (f filter) read() ([]byte, error) {
	for {
		char, err := f.in.read()
		if err == io.EOF {
			return f.dump(), err
		}
		f.buf = append(f.buf, char...)
		if f.isSecure() {
			f.clear()
			return f.mask, err
		} else if f.canDump() {
			return f.dump(), err
		}
	}
}

func (f filter) isSecure() bool {
	return string(f.buf) == f.str
}

func (f filter) canDump() bool {
	return !strings.HasPrefix(f.str, string(f.buf))
}

func (f filter) dump() []byte {
	chars := f.buf
	f.clear()
	return chars
}

func (f filter) clear() {
	f.buf = []byte{}
}

type writer struct {
	in  readable
	out io.Writer
}

func (w writer) run() {
	for {
		char, err := w.in.read()
		if err == io.EOF {
			break
		} else if err != nil {
			log.Fatalf("Error reading from input: %s", err)
		} else {
			w.out.Write(char)
		}
	}
}

func env(key string) string {
  return os.Getenv(key)
}

func strs() []string {
	strs := os.Args[1:]
  strs = mapStrs(strs, env)
	strs = choose(strs, func(str string) bool { return len(str) > 2 })
	strs = append(strs, mapStrs(strs, unescape)...)
	strs = uniq(strs)
	sort.Sort(sort.Reverse(byLength(strs)))
	return strs
}

func filtered(i io.Reader, o io.Writer, strs []string, mask string) writer {
	in := readable(stdin{in: i})
	for _, str := range strs {
		in = filter{in: in, str: str, mask: []byte(mask)}
	}
	return writer{out: o, in: in}
}

func main() {
	filtered(os.Stdin, os.Stdout, strs(), "[secure]").run()
}

File: filter.rs
use std::env;
use std::io::{self, Read};
use std::str;
use std::process::Command;

const MASK: &'static str = "[secure]";
const MIN_LEN: usize = 1;

trait Reader {
    fn read(&mut self) -> Option<String>;
}

struct Stdin {
    input: Box<std::io::Read>
}

impl Reader for Stdin {
    fn read(&mut self) -> Option<String> {
        let mut buffer = [0];
        let result = self.input.read(&mut buffer);

        if result.unwrap() == 0 {
            return None;
        } else {
            return Some(str::from_utf8(&buffer).unwrap().to_string());
        }
    }
}

struct Filter {
    input: Box<(Reader)>,
    string: String,
    buffer: String
}

impl Reader for Filter {
    fn read(&mut self) -> Option<String> {
        match self.input.read() {
            Some(c) => {
                self.buffer.push_str(&c);
                return Some(self.eval())
            },
            None => return None
        }
    }
}

impl Filter {
    fn eval(&mut self) -> String {
        if self.string == self.buffer {
            return self.mask();
        } else if !self.string.starts_with(&self.buffer) {
            return self.flush();
        } else {
            return "".to_string();
        }
    }

    fn mask(&mut self) -> String {
        self.buffer.clear();
        return MASK.to_string();
    }

    fn flush(&mut self) -> String {
        let str = self.buffer.clone();
        self.buffer.clear();
        return str
    }
}

fn unescape(cmd: &String) -> String {
    let out = Command::new("echo").arg(cmd).output().unwrap().stdout;
    return str::from_utf8(&out).unwrap().trim_right().to_string()
}

fn var(key: &String) -> String {
    env::var(key).unwrap_or("".to_string())
}

fn vars(keys: Vec<String>) -> Vec<String> {
    let mut strs: Vec<String> = keys.iter().map(|s| var(s)).collect();
    let mut escd: Vec<String> = strs.iter().map(|s| unescape(s)).collect();

    strs.append(&mut escd);
    strs.retain(|s| s.len() >= MIN_LEN);
    strs.sort();
    strs.dedup();
    strs.sort_by_key(|s| s.len());
    strs.reverse();

    return strs;
}

fn filter(stdin: Box<io::Read>, args: Vec<String>) -> Box<Reader> {
    let stdin: Box<Reader> = Box::new(Stdin { input: stdin });
    return vars(args).iter().fold(stdin, |input, arg| {
        Box::new(Filter { input: input, string: arg.to_string(), buffer: "".to_string() })
    });
}

fn run<'a>(input: Box<io::Read>, output: &'a mut io::Write, args: Vec<String>) {
    let mut filter = filter(input, args);
    while let Some(chars) = filter.read() {
        output.write(&chars.into_bytes()).unwrap();
        output.flush().unwrap();
    }
}

fn args() -> Vec<String> {
    let mut args: Vec<String> = env::args().collect();
    args.remove(0);
    return args;
}

fn main() {
    run(Box::new(io::stdin()), &mut io::stdout(), args());
}

#[cfg(test)]
mod test {
    use run;
    use std::io::Cursor;
    use std::env;

    #[test]
    fn test_foo() {
        env::set_var("a", "two");
        let args = vec!["a".to_string()];

        // http://stackoverflow.com/questions/28370126/how-can-i-test-stdin-and-stdout
        let input = Box::new(Cursor::new(&b"one two three"[..]));
        let mut output = Cursor::new(Vec::new());

        run(input, &mut output, args);

        let stdout = String::from_utf8(output.into_inner()).expect("Not UTF-8");
        assert_eq!("one [secure] three", stdout);
    }
}
